openadapt/privacy/providers/private_ai.py

```diff
@@ -0,0 +1,198 @@
+"""A Module for Private AI Scrubbing Provider."""
+
+from io import BytesIO
+from typing import List
+import base64
+
+from loguru import logger
+from PIL import Image
+import requests
+
+from openadapt import config
+from openadapt.privacy.base import Modality, ScrubbingProvider, TextScrubbingMixin
+from openadapt.privacy.providers import ScrubProvider
+
+BASE64_URL = "https://api.private-ai.com/deid/v3/process/files/base64"
+FILES_DIR = "assets/"
+HEADER_CONTENT_TYPE = "application/json"
+IMAGE_CONTENT_TYPE = "image/png"
+PDF_CONTENT_TYPE = "application/pdf"
+TEMP_IMAGEFILE_NAME = "temp_image_to_scrub.png"
+TEXT_URL = "https://api.private-ai.com/deid/v3/process/text"
+
+
+class PrivateAIScrubbingProvider(
+    ScrubProvider, ScrubbingProvider, TextScrubbingMixin
+):  # pylint: disable=abstract-method
+    """A Class for Private AI Scrubbing Provider."""
+
+    name: str = ScrubProvider.PRIVATE_AI
+    capabilities: List[Modality] = [Modality.TEXT, Modality.PIL_IMAGE, Modality.PDF]
+
+    def scrub_text(self, text: str, is_separated: bool = False) -> str:
+        """Scrub the text of all PII/PHI.
+
+        Args:
+            text (str): Text to be redacted
+            is_separated (bool): Whether the text is separated with special characters
+
+        Returns:
+            str: redacted text
+        """
+        payload = {
+            "text": [text],
+            "link_batch": False,
+            "entity_detection": {
+                "accuracy": "high",
+                "return_entity": True,
+            },
+            "processed_text": {
+                "type": "MARKER",
+                "pattern": "[UNIQUE_NUMBERED_ENTITY_TYPE]",
+            },
+        }
+
+        headers = {
+            "Content-Type": HEADER_CONTENT_TYPE,
+            "X-API-KEY": config.PRIVATE_AI_API_KEY,
+        }
+
+        response = requests.post(TEXT_URL, json=payload, headers=headers)
+
+        data = response.json()
+        logger.debug(f"{data=}")
+
+        # According to the PrivateAI API documentation,
+        # https://docs.private-ai.com/reference/latest/operation/process_text_v3_process_text_post/
+        # the response is a list of dicts when there is no error/issue in the request
+        # else it is a dict with a key "detail" containing the error message
+
+        if type(data) is dict and "detail" in data:
+            raise ValueError(data.get("detail"))
+
+        redacted_text = data[0].get("processed_text")
```

Comment:
Looks like this is raising an error in https://github.com/OpenAdaptAI/OpenAdapt/actions/runs/6609799609/job/17950542911?pr=486 :

```
==================================== ERRORS ====================================
_ ERROR collecting tests/openadapt/privacy/providers/test_private_ai_scrub.py __
tests/openadapt/privacy/providers/test_private_ai_scrub.py:18: in <module>
    scrub.scrub_text("hello Bob smith")
openadapt/privacy/providers/private_ai.py:73: in scrub_text
    redacted_text = data[0].get("processed_text")
E   KeyError: 0
------------------------------- Captured stderr --------------------------------
2023-10-23 07:05:11.403 | DEBUG    | openadapt.privacy.providers.private_ai:scrub_text:63 - data={'statusCode': 401, 'message': 'Access denied due to invalid subscription key. Make sure to provide a valid key for an active subscription.'}
=============================== warnings summary ===============================
../../../../Library/Caches/pypoetry/virtualenvs/openadapt-NqG5nKAR-py3.10/lib/python3.10/site-packages/onnxruntime/capi/_pybind_state.py:28
  /Users/runner/Library/Caches/pypoetry/virtualenvs/openadapt-NqG5nKAR-py3.10/lib/python3.10/site-packages/onnxruntime/capi/_pybind_state.py:28: DeprecationWarning: invalid escape sequence '\S'
    "(other than %SystemRoot%\System32), "

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
ERROR tests/openadapt/privacy/providers/test_private_ai_scrub.py - KeyError: 0
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
========================= 1 warning, 1 error in 18.57s =========================
```
It looks like a few things need to happen here:
1. Add a PrivateAI API key to the repo's environment variables. @KrishPatel13 can you please remind me whether we registered a key for OpenAdapt?
2. Fix the code here to handle this scenario
3. Modify the test not to run if the API key is not defined. It looks like this was implemented (https://github.com/OpenAdaptAI/OpenAdapt/pull/486/files#diff-2b5c61fcc6e55fe8105e356530deb8560174d69b8c1066cf9d62676f801aee0bR22) but is not functioning correctly.